packages/uploadthing/src/internal/handler.ts

```diff
@@ -488,6 +471,55 @@ export const buildRequestHandler = <TRouter extends FileRouter>(
   };
 };
 
+function resolveCallbackUrl(opts: {
+  config: RouterWithConfig<FileRouter>["config"];
+  req: RequestLike;
+  url: URL;
+}): URL {
+  let callbackUrl = opts.url;
+  if (opts.config?.callbackUrl) {
+    callbackUrl = getFullApiUrl(opts.config.callbackUrl);
+  } else if (process.env.UPLOADTHING_URL) {
+    callbackUrl = getFullApiUrl(process.env.UPLOADTHING_URL);
+  }
+
+  if (
+    process.env.NODE_ENV !== "production" ||
+    !callbackUrl.host.includes("localhost")
+  ) {
+    return callbackUrl;
+  }
+
+  // Production builds have to have a public URL so UT can send webhook
+  // Parse the URL from the headers
+  let parsedFromHeaders = (
+    getHeader(opts.req, "origin") ??
+    getHeader(opts.req, "referer") ??
+    getHeader(opts.req, "host") ??
+    getHeader(opts.req, "x-forwarded-host")
+  )?.toString();
+
+  if (parsedFromHeaders && !parsedFromHeaders.includes("http")) {
+    parsedFromHeaders =
+      (getHeader(opts.req, "x-forwarded-proto") ?? "https").toString() +
```

Comment:
L503-507 is a bit hard to parse. Mind breaking this out above, then using a template string?